cluster_pack/uploader.py

```diff
@@ -309,11 +340,29 @@ def _upload_env_from_venv(
         if not resolved_fs.exists(dir):
             resolved_fs.mkdir(dir)
         _logger.info(f'Uploading env at {local_package_path} to {package_path}')
-        resolved_fs.put(local_package_path, package_path)
+
+        if not _try_put_atomic(resolved_fs,
+                               local_package_path,
+                               package_path,
+                               lambda path: _is_archive_up_to_date(path, reqs, resolved_fs)):
+            raise FileExistsError(
+                f"Package with different version exists at {package_path}"
+            )
 
         _dump_archive_metadata(package_path, reqs, resolved_fs)
 
 
+def _check_pex_version(resolved_fs: filesystem.EnhancedFileSystem,
+                       local_package_path: str,
+                       remote_package_path: str) -> bool:
+    with tempfile.TemporaryDirectory() as tempdir:
+        local_copy_path = os.path.join(tempdir, os.path.basename(local_package_path))
+        resolved_fs.get(remote_package_path, local_copy_path)
+        into_at_destination = PexInfo.from_pex(local_copy_path)
+        info_to_upload = PexInfo.from_pex(local_package_path)
+    return info_to_upload.code_hash == into_at_destination.code_hash

```

Comment:
It would have been nicer to extract such changes that are not part of the ticket in another PR but LGTM